File: trinity-nexus/lang/libs/ruby/trinity_vsa/lib/trinity_vsa.rb
# frozen_string_literal: true

# Trinity VSA - Vector Symbolic Architecture with balanced ternary arithmetic
module TrinityVSA
  # Dense trit vector
  class TritVector
    attr_reader :data, :dim

    def initialize(data)
      @data = data
      @dim = data.length
    end

    def self.zeros(dim)
      new(Array.new(dim, 0))
    end

    def self.random(dim, seed: nil)
      srand(seed) if seed
      new(Array.new(dim) { rand(3) - 1 })
    end

    def [](i)
      @data[i]
    end

    def []=(i, v)
      @data[i] = v
    end

    def nnz
      @data.count { |x| x != 0 }
    end

    def sparsity
      1.0 - nnz.to_f / dim
    end

    def negate!
      @data.map! { |x| -x }
    end

    def clone
      TritVector.new(@data.dup)
    end
  end

  class << self
    # Bind two vectors (element-wise multiplication)
    def bind(a, b)
      raise 'Dimension mismatch' unless a.dim == b.dim
      TritVector.new(a.data.zip(b.data).map { |x, y| x * y })
    end

    # Unbind (inverse of bind)
    def unbind(a, b)
      bind(a, b)
    end

    # Bundle vectors via majority voting
    def bundle(vectors)
      raise 'Empty vector list' if vectors.empty?
      dim = vectors[0].dim
      result = Array.new(dim) do |i|
        sum = vectors.sum { |v| v[i] }
        sum > 0 ? 1 : (sum < 0 ? -1 : 0)
      end
      TritVector.new(result)
    end

    # Circular permutation
    def permute(v, shift)
      TritVector.new(v.data.rotate(-shift))
    end

    # Dot product
    def dot(a, b)
      raise 'Dimension mismatch' unless a.dim == b.dim
      a.data.zip(b.data).sum { |x, y| x * y }
    end

    # Cosine similarity
    def similarity(a, b)
      d = dot(a, b).to_f
      norm_a = Math.sqrt(dot(a, a))
      norm_b = Math.sqrt(dot(b, b))
      return 0.0 if norm_a.zero? || norm_b.zero?
      d / (norm_a * norm_b)
    end

    # Hamming distance
    def hamming_distance(a, b)
      raise 'Dimension mismatch' unless a.dim == b.dim
      a.data.zip(b.data).count { |x, y| x != y }
    end
  end
end

File: trinity-nexus/lang/libs/elixir/trinity_vsa/lib/trinity_vsa.ex
defmodule TrinityVsa do
  @moduledoc """
  Trinity VSA - Vector Symbolic Architecture with balanced ternary arithmetic
  """

  @doc "Create zero vector"
  def zeros(dim), do: List.duplicate(0, dim)

  @doc "Create random vector"
  def random(dim, seed \\ nil) do
    if seed, do: :rand.seed(:exsss, {seed, seed, seed})
    Enum.map(1..dim, fn _ -> :rand.uniform(3) - 2 end)
  end

  @doc "Bind two vectors (element-wise multiplication)"
  def bind(a, b) when length(a) == length(b) do
    Enum.zip_with(a, b, &(&1 * &2))
  end

  @doc "Unbind (inverse of bind)"
  def unbind(a, b), do: bind(a, b)

  @doc "Bundle vectors via majority voting"
  def bundle([]), do: raise("Empty vector list")
  def bundle(vectors) do
    dim = length(hd(vectors))
    Enum.map(0..(dim - 1), fn i ->
      sum = Enum.sum(Enum.map(vectors, &Enum.at(&1, i)))
      cond do
        sum > 0 -> 1
        sum < 0 -> -1
        true -> 0
      end
    end)
  end

  @doc "Circular permutation"
  def permute(v, shift) do
    dim = length(v)
    {left, right} = Enum.split(v, rem(dim - shift, dim))
    right ++ left
  end

  @doc "Dot product"
  def dot(a, b) when length(a) == length(b) do
    Enum.zip_with(a, b, &(&1 * &2)) |> Enum.sum()
  end

  @doc "Cosine similarity"
  def similarity(a, b) do
    d = dot(a, b)
    norm_a = :math.sqrt(dot(a, a))
    norm_b = :math.sqrt(dot(b, b))
    if norm_a == 0 or norm_b == 0, do: 0.0, else: d / (norm_a * norm_b)
  end

  @doc "Hamming distance"
  def hamming_distance(a, b) when length(a) == length(b) do
    Enum.zip_with(a, b, fn x, y -> if x != y, do: 1, else: 0 end) |> Enum.sum()
  end
end
